### ISSUE TITLE
Strange behavior of kv_store

### ISSUE THREAD

**Issue by superzrx** [apache/incubator-mxnet#1556]
Hi guys,
    It seems that different type of kv_store will lead to different training curve. 
    The only difference between following runs is the type of kv_store. Each cell contains Train/Val accuracy shown at the end of epoch.

| epoch | local-run0 | local-run1 | local-run2 | device-run0 | device-run1 |
| --- | --- | --- | --- | --- | --- |
| 0 | 0.119/0.250 | 0.126/0.242 | 0.123/0.246 | 0.124/0.262 | 0.119/0.250 |
| 1 | 0.275/0.352 | 0.281/0.366 | 0.280/0.362 | 0.290/0.379 | 0.286/0.376 |
| 2 |  | 0.338/0.396 | 0.337/0.395 | 0.372/0.434 | 0.368/0.435 |
| 3 |  | 0.364/0.421 |  | 0.421/0.480 | 0.418/0.474 |
| 4 |  | 0.380/0.440 |  | 0.453/0.497 |  |

 As shown in the table above, kv_store with type device tends to learn faster than local case.
Is this normal? @mli

**Comment by mli** [comment]
the reason that different local runs have difference results are due to

1. random data crop, weight inited with random seed
2. cudnn give random results each time

local and device are quite difference is possibly due to weight initiazation. local is inited on cpu (by glibc library), while devcie is inited on gpu (by cuda)

**Comment by mli** [comment]
the difference should be much smaller if you run more epochs.

**Comment by superzrx** [comment]
@mli 
I run two resnet-50 experiment with different kv-store type and on two machines while other configuration keeps same.
After 30 epochs the one with 'local' ( also has /2 on 1x1) gets about 50% while 'device' reaches 65%. 

Could you show me some links to how device works? It feels like has a smaller equivalent learning rate

**Comment by mli** [comment]
i added a test case for training with various kvstores #1826 , they should give very similar results if we try to eliminate randomness. however, i observed that cpu and gpu give quite different results even if the inputs are identical. 

- local: both gradients are averaged and parameters are updated on cpu
- device: both are done at gpu

so if you 

- disable cudnn
- disable all randomness in data iterator,  for example
```python
train = mx.io.ImageRecordIter(
        path_imgrec = "data/cifar/train.rec",
        mean_img    = "data/cifar/cifar_mean.bin",
        data_shape  = input_shape,
        batch_size  = batch_size,
        rand_crop   = False,
        rand_mirror = False,
        shuffle     = False,
        round_batch = False,
        num_parts   = num_parts,
        part_index  = part_index)
    val = mx.io.ImageRecordIter(
        path_imgrec = "data/cifar/test.rec",
        mean_img    = "data/cifar/cifar_mean.bin",
        rand_crop   = False,
        rand_mirror = False,
        shuffle     = False,
        round_batch = False,
        data_shape  = input_shape,
        batch_size  = batch_size)
    return (train, val)
```
- increase the batch size to reduce the number of iterations

both local and device should give similar results on epoch 1.

**Comment by superzrx** [comment]
@mli local gives similar results compare to other libraries like torch
maybe device has some problem
in torch nccl is used to speed up gpu communication, do we have any plan to include it?

**Comment by szha** [comment]
I'm closing this issue due to lack of activity. Feel free to ping me to reopen if this is still an active issue.
